### DISCUSSION TITLE
[C# feature request] Default value for Dictionary<>

### DISCUSSION THREAD

**Discussion by amshaikh** [dotnet/runtime#42383] [General]
It would really be handy if the user of the Dictionary<> class be able to set a default value which should be returned if the key is not found in the dictionary.

eg: (Wished code)

Dictionary<string, int> TotalMarks = new  Dictionary<string, int>();
TotalMarks.DefaultValue = 0;
Console.WriteLine(TotalMarks["Tom"]); //should work --no need to check whether the key exists or not.

**Comment by amshaikh** [comment]
Use the following wrapper:

    public class Map<TKey,TValue> : Dictionary<TKey,TValue>
    {
        public TValue this [TKey key]
        {
            get
            {
                if (base.ContainsKey(key) || !isDefaultValueSet)
                {
                    return base[key];
                }
                else
                {
                    return DefaultValue;
                }
            }
            set { base[key] = value; }
        }

        private bool isDefaultValueSet = false;
        private TValue _DefaultValue;
        public TValue DefaultValue
        {
            get { return _DefaultValue; }
            set { _DefaultValue = value; isDefaultValueSet = true; }
        }
    }

**Comment by YairHalberstadt** [comment]
This is a request for dotnet/runtime

**Comment by GrabYourPitchforks** [comment]
I think @amshaikh provided the solution (though for efficiency the caller should use `TryGetValue` instead of `ContainsKey`). Is there anything else to discuss here?

**Comment by sharwell** [comment]
This functionality is provided through the `GetValueOrDefault` extension methods:
https://github.com/dotnet/runtime/blob/839b1343fbad61526fedcaa722f1ae48b298c543/src/libraries/System.Collections/src/System/Collections/Generic/CollectionExtensions.cs#L10-L24

**Reply by GrabYourPitchforks** [reply]
API refs:
- https://apisof.net/catalog/System.Collections.Generic.CollectionExtensions
- https://docs.microsoft.com/dotnet/api/system.collections.generic.collectionextensions
